File: IRCClient/src/main/java/net/darkmeow/irc/client/network/handle/HandleClientEncryption.java
package net.darkmeow.irc.client.network.handle;

import io.netty.channel.ChannelHandlerContext;
import io.netty.channel.ChannelInboundHandlerAdapter;
import io.netty.handler.codec.LengthFieldBasedFrameDecoder;
import net.darkmeow.irc.client.network.IRCClientConnection;
import net.darkmeow.irc.client.network.handle.encryption.HandleClientEncryptionInbound;
import net.darkmeow.irc.client.network.handle.encryption.HandleClientEncryptionOutbound;
import net.darkmeow.irc.network.IRCNetworkBaseConfig;

public class HandleClientEncryption extends ChannelInboundHandlerAdapter {

    public final IRCClientConnection client;

    public HandleClientEncryption(IRCClientConnection client) {
        this.client = client;
    }

    @Override
    public void channelActive(ChannelHandlerContext ctx) throws Exception {
        ctx.pipeline().addBefore("BaseEncryption", "BaseFrameDecoder", new LengthFieldBasedFrameDecoder(
            IRCNetworkBaseConfig.MAX_FRAME_LENGTH,
            0,
            4,
            0,
            4
        ));
        ctx.pipeline().addAfter(
            "BaseFrameDecoder",
            "BaseEncryptionInbound",
            new HandleClientEncryptionInbound(client)
        );
        ctx.pipeline().addAfter(
            "BaseFrameDecoder",
            "BaseEncryptionOutbound",
            new HandleClientEncryptionOutbound(client)
        );

        super.channelActive(ctx);
    }

}

File: IRCServer/src/main/kotlin/net/darkmeow/irc/network/handles/HandleClientEncryption.kt
package net.darkmeow.irc.network.handles

import io.netty.channel.ChannelHandlerContext
import io.netty.channel.ChannelInboundHandlerAdapter
import io.netty.handler.codec.LengthFieldBasedFrameDecoder
import net.darkmeow.irc.network.IRCNetworkBaseConfig
import net.darkmeow.irc.network.NetworkManager
import net.darkmeow.irc.network.handles.encryption.HandleClientEncryptionInbound
import net.darkmeow.irc.network.handles.encryption.HandleClientEncryptionOutbound

class HandleClientEncryption(private val manager: NetworkManager): ChannelInboundHandlerAdapter() {

    override fun channelActive(ctx: ChannelHandlerContext) {
        ctx.pipeline().addBefore(
            "BaseEncryption",
            "BaseFrameDecoder",
            LengthFieldBasedFrameDecoder(
                IRCNetworkBaseConfig.MAX_FRAME_LENGTH,
                0,
                4,
                0,
                4
            )
        )
        ctx.pipeline().addAfter(
            "BaseFrameDecoder",
            "BaseEncryptionInbound",
            HandleClientEncryptionInbound(manager)
        )
        ctx.pipeline().addAfter(
            "BaseFrameDecoder",
            "BaseEncryptionOutbound",
            HandleClientEncryptionOutbound(manager)
        )

        super.channelActive(ctx)
    }

}